pkg/agent/watch.go

```diff
@@ -101,19 +100,8 @@ func startVMWatcher(
 				}
 			},
 			UpdateFunc: func(oldVM, newVM *vmapi.VirtualMachine) {
-				oldMetrics := metricsForVM(oldVM, nodeName)
-				newMetrics := metricsForVM(newVM, nodeName)
-
-				// delete old metric label sets that are no longer present
-				for k := range oldMetrics {
-					if _, ok := newMetrics[k]; !ok {
-						perVMMetrics.vmResources.DeleteLabelValues(k.labelValues()...)
-					}
-				}
-				// set all current metric values
-				for k, v := range newMetrics {
-					perVMMetrics.vmResources.WithLabelValues(k.labelValues()...).Set(v)
-				}
+				deleteMetricsfromVM(oldVM, nodeName, &perVMMetrics)
+				setMetricsFromVM(newVM, nodeName, &perVMMetrics)
```

Comment:
How can we have race condition here via "delete+add"? My understanding was perVMMetrics is being used only by one goroutine at the time, so we don't have any synchronization for it.

Reply:
the race would be:

1. watcher deletes VM metrics
2. metrics get scraped externally
3. watcher adds back VM metrics

Reply:
I see, but this race should have existed in the previous version as well:

https://github.com/neondatabase/autoscaling/blob/main/pkg/agent/watch.go#L107-L116

Reply:
In some form, yes - however: the regression is that before, the race only existed if the labels changed (because we currently don't delete the label sets if they're present after the change as well). With this change, the race could exist on any update event for a VM, 

Note that because the labels only change if the endpoint ID changes (and we're _probably_ fetching metrics by endpoint ID rather than compute ID), the discontinuity there is probably not possible to notice because endpoint IDs are changed at most once, going from `""` to the assigned endpoint ID for the compute.

Reply:
Makes sense, though it's tricky. I'd make functions for such operations (such as set, delete, and update) to make sure we're handling everything in the right way in one place.

Reply:
These are on the other hand _metrics_. There are 14400 datapoints every day and individual ones aren't terribly interesting, starting from the fact the panels we draw these to are just some hundreds of pixels wide and each pixel can represent a single datapoint.

Scraping can also happen bitween any given two `Set`s so there's no real atomicity anyway, but that's business as usual.

We aren't interested in metrics that don't have endpointID, because that would mean pooled compute.

That being said, I didn't think much further the update case. What are the concrete cases where we actually would need to delete something instead of just setting?

Reply:
> There are 14400 datapoints every day and individual ones aren't terribly interesting, starting from the fact the panels we draw these to are just some hundreds of pixels wide and each pixel can represent a single datapoint

Yeah, that's fair - I do worry about cases where e.g. a user is looking at the last 5 minutes of data and there's a gap, but perhaps this isn't really an issue in practice.

> That being said, I didn't think much further the update case. What are the concrete cases where we actually would need to delete something instead of just setting?

I think deletion would happen in one of two cases:

1. the VM switches to a different node (either by migration or restart)
2. the endpoint ID label gets set on the VM, going from metrics with `endpoint_id=""` to `endpoint_id!=""`